storage/redis/peer_store.go

```diff
@@ -0,0 +1,828 @@
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, group := range ps.groups() {
+		// list all infohashes in the group
+		infohashesList, err := redis.Strings(conn.Do("HKEYS", group))
+		if err != nil {
+			return err
+		}
+
+		for _, ihStr := range infohashesList {
+			isSeeder := len(ihStr) > 5 && ihStr[5:6] == "S"
+
+			// list all (peer, timeout) pairs for the ih
+			ihList, err := redis.Strings(conn.Do("HGETALL", ihStr))
+			if err != nil {
+				return err
+			}
+
+			var pk serializedPeer
+			var removedPeerCount int64
+			for index, ihField := range ihList {
+				if index%2 == 1 { // value
+					mtime, err := strconv.ParseInt(ihField, 10, 64)
+					if err != nil {
+						return err
+					}
+					if mtime <= cutoffUnix {
+						log.Debug("storage: deleting peer", log.Fields{
+							"Peer": decodePeerKey(pk).String(),
+						})
+						ret, err := redis.Int64(conn.Do("HDEL", ihStr, pk))
+						if err != nil {
+							return err
+						}
+
+						removedPeerCount += ret
+					}
+				} else { // key
+					pk = serializedPeer([]byte(ihField))
+				}
+			}
+			// DECR seeder/leecher counter
+			decrCounter := ps.leecherCountKey(group)
+			if isSeeder {
+				decrCounter = ps.seederCountKey(group)
+			}
+			if removedPeerCount > 0 {
+				if _, err := conn.Do("DECRBY", decrCounter, removedPeerCount); err != nil {
+					return err
+				}
+			}
+
+			// use WATCH to avoid race condition
+			// https://redis.io/topics/transactions
+			_, err = conn.Do("WATCH", ihStr)
+			if err != nil {
+				return err
+			}
+			ihLen, err := redis.Int64(conn.Do("HLEN", ihStr))
+			if err != nil {
+				return err
+			}
+			if ihLen == 0 {
+				// Empty hashes are not shown among existing keys,
+				// in other words, it's removed automatically after `HDEL` the last field.
+				//_, err := conn.Do("DEL", ihStr)
+
+				conn.Send("MULTI")
+				conn.Send("HDEL", group, ihStr)
+				if isSeeder {
+					conn.Send("DECR", ps.infohashCountKey(group))
+				}
+				_, err = redis.Values(conn.Do("EXEC"))
+				if err != nil && err != redis.ErrNil {
+					log.Error("storage: Redis EXEC failure", log.Fields{
+						"group":    group,
+						"infohash": ihStr,
+						"error":    err,
+					})
+				}
+			} else {
+				if _, err = conn.Do("UNWATCH"); err != nil && err != redis.ErrNil {
+					log.Error("storage: Redis UNWATCH failure", log.Fields{"error": err})
+				}
+			}
+		}
+	}
+
+	duration := float64(time.Since(start).Nanoseconds()) / float64(time.Millisecond)
+	log.Debug("storage: recordGCDuration", log.Fields{"timeTaken(ms)": duration})
+	storage.PromGCDurationMilliseconds.Observe(duration)
+
+	return nil
+}
+
+func (ps *peerStore) Stop() stop.Result {
+	c := make(stop.Channel)
+	go func() {
+		close(ps.closed)
+		ps.wg.Wait()
+		// chihaya does not clear data in redis when exiting.
+		// chihaya keys have prefix `IPv{4,6}_`.
```

Comment:
Can you turn this into an info-level log message?

Reply:
sure